---
title: "Part 2"
output: html_notebook
---
```{r setup, eval = FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

In this exercise are asked to compute the likelihood of the sequencing data based on

* genotype likelihoods as infered from the quality score
* allele frequencies provided by the NGSadmix model, which are in turn given by the admixture proportions and the ancestral allele frequencies

Load some libraries
```{r}
library("ggplot2")
theme_set(theme_bw())
library("tidyr")
library("dplyr")
```

```{r, echo = F, include = F}
N <- 100
M <- 5e4
```

Load $Q_{i}$ (admixture proportions matrix) and $F_{j}$ (ancestral allele frequencies).
Load population panel.

```{r}
fopt <- read.table(file = "assign3.fopt.gz") %>% head(M)
qopt <- read.table(file = "assign3.qopt") %>% head(N)
qopt$id <- 1:nrow(qopt)
pop_info <- read.table(file = "pop.info") %>% head(N)
individuals <- pop_info$V2 %>% as.character %>% unique
qopt <- cbind(qopt, pop = pop_info$V1, individual = pop_info$V2)
```

Load genotype likelihoods $P(X_{ij}| G_{ij})$. Read as likelihood of genotype G for individual i in site j, given the observed data X.
Organize in a tidy format (indexed by likelihoods).
```{r}
input <- read.table(file = "input.gz", header = T, nrows = M)
markers <- as.character(input$marker) %>% unique
rownames(fopt) <- markers
res <- input[,4:303]
result <- list()
k <- 1
for(i in seq(1, N*3, 3)) {
  result[[k]] <- c(res[,i], res[,i+1], res[,i+2])
  k <- k + 1
}
result <- do.call(cbind, result)
input <- select(input, marker, allele1, allele2)
genotypes <- data.frame(marker = rep(input$marker, each = 3),
                        genotype = rep(0:2, times = nrow(input)))
colnames(result) <- pop_info$V2 %>% as.character

df <- full_join(input, genotypes, by = "marker")
df$marker <- factor(df$marker, levels = markers)
df <- df%>% arrange(-genotype, marker)
input <- cbind(df, result)
input <- gather(input, individual, likelihood, -marker, -allele1, -allele2, -genotype)
input$loglik <- log(input$likelihood)
```

Visualize the contribution of each ancestral to the current populations
```{r}  
df <- qopt %>%
  group_by(pop) %>%
  summarise(V1 = mean(V1),
            V2 = mean(V2),
            V3 = mean(V3)) %>%
  gather(ancestral, proportion, -pop)

p <- ggplot(data = df,
       mapping = aes(x = pop, y = proportion, fill = ancestral)) + 
  geom_bar(stat = "identity") +
  scale_fill_manual(values = c("black", "yellow", "#ffff99"))
print(p)
```



```{r, echo = F, include = F}
max <- df %>% group_by(ancestral) %>% summarise(max = which.max(proportion)) %>% .$max
levels(df$pop)[max]
```
Given this results, we state the first ancestral population is made up by the African inheritance, the second by the European, and the third by the Asian.

Rename V1:V3 accordingly in the data
```{r}
ancestries <- c("African", "European", "Asian")
df$ancestral <- df$ancestral %>%
  gsub(pattern = "V1", replacement =  ancestries[1]) %>%
  gsub(pattern = "V2", replacement = ancestries[2]) %>%
  gsub(pattern = "V3", replacement = ancestries[3])
colnames(qopt)[1:3] <- ancestries
```

Afro American individuals are 
```{r}
(df %>% filter(pop == "ASW", ancestral == "Asian") %>% .$proportion) * 100 
```
% Asian on average.

Visualize the Asian proportion distribution in the Afro American individuals.
```{r}
filter(qopt, pop == "ASW") %>% ggplot(mapping = aes(x = Asian, fill = pop)) + geom_histogram() +
  scale_y_continuous(breaks = seq(0, 10, 2) ) + labs(y = "ASW individuals count") + scale_fill_manual(values = "yellow") + guides(fill = F)
```

Admixture proportions under the alternative model
```{r}
alt <- filter(qopt, individual == "NA19818") %>% select(African:Asian)
alt
```
Admixture proportions under the null model. The Asian contribution is added to the European.
```{r}
null <- alt
null$European <- null$Asian + null$European
null$Asian <- 0
```

A function to compute $h_{ij}$ by taking the weighted average of the ancestral frequencies and the admixture proportions 
The function computes $h$ for all sites in a single individual.
```{r}
compute_h <- function(proportions, frequencies, ind) {
  prop <- filter(proportions, individual == ind) %>%
    select(African:Asian) %>%
    unlist
 
  freq <- as.matrix(frequencies)

  h <- (prop * t(freq)) %>%
    colSums %>%
    as.matrix(nrow = nrow(frequencies), ncol = 1)
  
  colnames(h) <- ind
  rownames(h) <- rownames(frequencies)
  return(h)
}
```

A function to implement the likelihood model. Computes the genotype probabilities assumming Hardy Weinberg equilibrium for fiven F and Q and multiplies that with the genotype likelihoods.
```{r}
likelihood_model <- function(likelihoods, h, ind=NULL, verbose=F) {
  sites <- 1:M
  genotypes <- 0:2
  markers <- likelihoods$marker %>% as.character %>% unique
  ancestries <- c("African", "European", "Asian")
  
  if(is.null(ind)) {
    print("Please supply individual id within h and likelihoods")
    return(NULL)
  }
  
  prob <- 0
  for(s in sites) {
    if(s %% (M/50) == 0 & verbose)
      print(s)
      
    acum <- 0
    for(g in genotypes) {
      # Compute likelihood of model given data by summing the log of 2 probabilities
      # the likelihood of the current genotype given the data (from the input.gz file)
      # the probability of the genotype given the model
      hfq <- ifelse(g == 2,
                    h[s,]**2,
                    ifelse(
                      g == 1,
                      2 * h[s,] * (1 - h[s,]),
                      (1-h[s,])**2
                      ))
      
      lik <- likelihoods %>% filter(individual == ind,
                                   marker == markers[s],
                                   genotype == g) %>% .$likelihood
      
      p <- hfq * lik
      acum <- acum + p
    }
          # This is a list of 3 numbers containing the log likelihoods for the 3 possible genotypes
          # If one of them is -Inf, the original probability was 0. Just treat this Infinites as 0
    prob <- prob + log(acum)
    
        
  }
  return(prob)
}
```

```{r, eval = F}
ind <- "NA19818"
h_alt <- compute_h(proportions = qopt, frequencies = fopt, ind = ind)
alt_result <- likelihood_model(
    likelihoods = input,
    h = h_alt,
    ind = ind)

h_null <- compute_h(proportions = null_qopt, frequencies = fopt, ind = ind)
null_result <- likelihood_model(
    likelihoods = input,
    h = h_null,
    ind = ind
    )
```




Set null model as special case of alternative model when Asian proportion is transfered to European proportion.
```{r}
null_qopt <- qopt
null_qopt$European <- null_qopt$European + null_qopt$Asian
null_qopt$Asian <- 0
```





```{r}
likelihood_ratio_test <- function(likelihoods, fopt, null_qopt, alt_qopt, ind, dfreedom = M + N, verbose = F) {
  # individuals <- null_qopt$individual
  
  if(filter(alt_qopt, individual == ind)$Asian == 0) {
    chi <- 0
    pvalue <- 1
  } else {
    h_alt <- compute_h(qopt, fopt, ind)
    h_null <- compute_h(null_qopt, fopt, ind)
    
    
    null_result <- likelihood_model(
      likelihoods = likelihoods,
      h = h_null,
      ind = ind
      )
    
    alt_result <- likelihood_model(
      likelihoods = likelihoods,
      h = h_alt,
      ind = ind,
      verbose = verbose
      )
    
    chi <- 2 * (alt_result - null_result)
    pvalue <- 1 - pchisq(q = chi, df = dfreedom)
    }
  return(list(chi = chi, pvalue = pvalue))
}
```


```{r}
ind <- "NA18550"
s <- sample(1:100, size = 1)
h_alt <- compute_h(proportions = qopt, frequencies = fopt, ind = ind)
h_null <- compute_h(proportions = null_qopt, frequencies = fopt, ind = ind)
h_alt <- h_alt[s,]
h_null <- h_null[s,]
lrt <- likelihood_ratio_test(input, fopt, null_qopt, qopt, ind = ind)
a <- filter(input, marker == markers[s], individual == ind) %>% .$likelihood
    
h_alt ** 2 * a[3] + (2 * h_alt * (1 - h_alt)) * a[2] + (1-h_alt)**2 * a[1]

h_null ** 2 * a[3] + (2 * h_null * (1 - h_null)) * a[2] + (1-h_null)**2 * a[1]
```


```{r}
lrt <- likelihood_ratio_test(input, fopt, null_qopt, qopt, ind = ind)
lrt <- likelihood_ratio_test(input, fopt, null_qopt, qopt, ind = "NA19818")
lrt <- likelihood_ratio_test(input, fopt, null_qopt, qopt, ind = "NA19625")
lrt <- likelihood_ratio_test(input, fopt, null_qopt, qopt, ind = "NA18546")

likelihood_ratio_test(input, fopt, null_qopt, qopt, ind = "NA19916")
```
```{r}
qopt[qopt$individual == "NA19818","Asian"] <- 0.0001
qopt[qopt$individual == "NA19818","European"] <- 1 - qopt[qopt$individual == "NA19818","African"]
likelihood_ratio_test(input, fopt, null_qopt, qopt, ind = "NA19818")
```

```{r}
lrt <- likelihood_ratio_test(input, fopt, null_qopt, qopt, ind = filter(null_qopt, pop == "ASW") %>% .$individual %>% as.character)
sum(lrt$pvalue < 0.05)
```


```{r}
## Admixture plot
res <- qopt[,1:3] < 0.001
qopt[,1:3][res] <- 0

long_qopt <- gather(arrange(qopt, -African, -European, -African), population, proportion, -id, -pop, -individual)
long_qopt$id <- factor(long_qopt$id, levels = unique(long_qopt$id))

q <- ggplot(data = long_qopt,
       mapping = aes(x = id, y = proportion, fill = population)) +
  geom_bar(stat = "identity", colour = "transparent",
           width = 1, show.legend = F) +
  theme_void() +
  theme(strip.background = element_blank(),
        strip.text.y = element_blank(),
        strip.text.x = element_blank()) +
  scale_y_continuous(expand = c(0, 0)) +
  scale_fill_manual(values = c("black", "yellow", "#ffff99"))
q
```

```{r}
library("knitr")
purl("part2.Rmd")
```

